### DISCUSSION TITLE
How to auto-reload language server (started in debug mode) when changing code

### DISCUSSION THREAD

**Discussion by movalz** [eclipse-langium/langium#1561] [Q&A]
Using VSCode I want to debug the language server with "hot code replacement" like I'm used to when developing with Xtext in Eclipse.

I have tried a configuration in launch.json to start the language server with a `"runtimeExecutable": "nodemon"`:
```
        {
            "name": "Run Extension with nodemon",
            "type": "node",
            "request": "launch",
            "runtimeExecutable": "nodemon",
            "runtimeArgs": [
                "--watch",
                "out",
                "--ext",
                "js,cjs",
                "--exec",
                "node --inspect /my-test-project/out/extension/main.cjs"
            ],
            "outFiles": [
                "/my-test-project/out/**/*.js"
            ],
            "sourceMaps": true,
            "restart": true
        }
```

When I...
1. start "npm run watch" and
2. start the launch config via "start debugging" I see nodemon output in the console (e.g. "starting 'node --inspect /my-test-project/out/extension/main.cjs'"), but it's followed by an error message ("Uncaught Error Error [ERR_REQUIRE_ESM]: require() of ES Module C:\my-test-project\node_modules\vscode\api.js from C:\my-test-project\out\extension\main.cjs not supported").

Is it possible to get something like the hot code replacement of Eclipse with Langium development with VSCode?

**Comment by msujew** [comment]
Hey @movalz,

good question, though I'm not sure how easily possible it is to perform hot code replacement in node.js/VS Code at all. The error you're running into specifically is due to the fact that you cannot run a VS Code extension on its own. You need to run a development host instance of VS Code and instruct it to load your extension (which is done in the default launch config). AFAIK VS Code doesn't support hot reloading of its extensions. Instead, the whole extension host needs to be restarted.

I believe this might rather be a question for the guys over at the [vscode-languageserver-node](https://github.com/microsoft/vscode-languageserver-node) or [vscode](https://github.com/microsoft/vscode) repo directly. This is probably out of scope of Langium.

Note that the way we do iterative manual testing on language servers is by having a watch script recompile the language server+extension on every change and then reloading the VS Code instance manually using the debug controlls:

![grafik](https://github.com/eclipse-langium/langium/assets/4377073/4daaf227-19e0-41cb-bc3b-5ba842992ce3)

**Reply by movalz** [reply]
thank you for the answer. So I will use the manual reload first.
